chain/consensus/mir/manager.go

```diff
@@ -186,20 +270,34 @@ func (m *Manager) Stop() {
 	log.With("miner", m.MirID).Infof("Mir manager shutting down")
 	defer log.With("miner", m.MirID).Info("Mir manager stopped")
 
-	if err := m.Wal.Close(); err != nil {
+	if err := m.WAL.Close(); err != nil {
 		log.Errorf("Could not close write-ahead log: %s", err)
 	}
 	log.Info("WAL closed")
 
 	m.Net.Stop()
 }
 
+// ReconfigureMirNode reconfigures the Mir node.
+func (m *Manager) ReconfigureMirNode(ctx context.Context, nodes map[t.NodeID]t.NodeAddress) error {
+	log.With("miner", m.MirID).Debug("Reconfiguring a Mir node")
+
+	if len(nodes) == 0 {
+		return fmt.Errorf("empty validator set")
+	}
+
+	go m.Net.Connect(ctx, nodes)
```

Comment:
Just to understand what is happening under the hood. We connect to all nodes (in case there is one we are not connected to in the new set), and we garbage collect the connections of the nodes that left, right? 

Why do we spawn different routines for this? Can this lead to a data race if Mir tries to send a message to a node before we have successfully connected to it (maybe there is an internal lock that handles this and makes this operation concurrent-friendly, but I wanted to double-check just in case).

Reply:
> Just to understand what is happening under the hood. We connect to all nodes (in case there is one we are not connected to in the new set), and we garbage collect the connections of the nodes that are left, right?

That's right.

> Why do we spawn different routines for this?

Mir's libp2p transport can try to establish connections with a node very long time. To not block this we spawn a new goroutine. We believe that adding a node is a rare operation.

@matejpavlovic can elaborate more about possible data race in Mir

Reply:
That makes sense as long as Mir message sender is blocked for the corresponding peer until the connection is established. Let's wait for @matejpavlovic input.

Reply:
Yes indeed, this actually is an issue, good catch. A node could try to send a message to a newly joined node that the Net module hasn't even tried to connect to yet, resulting in an error (which in the current implementation would even lead to a crash). In this concrete case it is pretty unlikely, but still possible.

The Net module needs to be aware of all the node IDs it can be asked to send messages to before any send event. To do that, maybe we could have the `Connect` method not run in a separate goroutine (i.e. just call `m.Net.Connect(ctx, nodes)`, without the `go`), but change its internal implementation to a non-blocking one that creates internal message buffers synchronously, but builds the network connections in the background.

Moreover, we should be more graceful about sending messages to unknown nodes. Simply dropping the message (and maybe logging an error) would be the right thing to do in my opinion.

Reply:
I created an issue in Mir for that.

Reply:
> The Net module needs to be aware of all the node IDs it can be asked to send messages to before any send event. To do that, maybe we could have the Connect method not run in a separate goroutine (i.e. just call m.Net.Connect(ctx, nodes), without the go), but change its internal implementation to a non-blocking one that creates internal message buffers synchronously, but builds the network connections in the background.

This would be a good fix in the current PR, and I think is a good idea for Mir to have a sanity-check as a library. Thanks!